File: src_Matlab/CalcSysVars.m
classdef CalcSysVars
    % CALCSYSVARS - A class that calculates the system variables:         
    %   time-span
    %   Initial state vector vars - lengths and angles subtended

    properties (SetAccess = private, GetAccess = protected)
        pInitial_, jInitial_,xTarget_,yTarget_;startTime_,endTime_,timeSteps_;
    end
    
    methods
        % Constructor
        function obj = CalcSysVars(coordinates,time_params)
            obj.pInitial_ = coordinates(1:2);
            obj.jInitial_ = reshape(coordinates(3:6)',[2,2]);
            obj.xTarget_ = coordinates(7);
            obj.yTarget_ = coordinates(8);
            obj.startTime_ = time_params(1);
            obj.endTime_ = time_params(2);
            obj.timeSteps_ = time_params(3);
        end

        function [x_init] = getInitialVars(obj)
            % Calculating initial length between the actuator joint and the platform joint at the starting position
            OB_i = zeros(2,2);
            Rot = zeros(2,2);
            Length_init = zeros(1,2);

            for i=1:2
                OB_i(:,i)=obj.pInitial_'-obj.jInitial_(:,i);
            end
            for i=1:2
                Rot(:,i)=OB_i(:,i)/norm(OB_i(:,i)); % vector / norm (vector)
                Length_init(:,i)=norm(OB_i(:,i)); % Gives the length of the Actuator arms
            end 
            phi_init=[atan2(Rot(2,1),Rot(1,1)) atan2(Rot(2,2),Rot(1,2))];
            x_init = [Length_init(1) 0 Length_init(2) 0 phi_init(1) 0 phi_init(2) 0 obj.xTarget_ obj.yTarget_];
        end

        function time_span = getTimeSpan(obj)
            N = (obj.endTime_ - obj.startTime_)/obj.timeSteps_;
            time_span = linspace(obj.startTime_,obj.endTime_,N);
        end
    end
end

File: src/CalcSysVars.cpp

// Include Files
#include "CalcSysVars.h"
#include "linspace.h"
#include "norm.h"
#include "coder_array.h"
#include <cmath>

// Function Definitions
//
// Calculating initial length between the actuator joint and the platform joint
// at the starting position
//
// Arguments    : double x_init[10]
// Return Type  : void
//
void CalcSysVars::getInitialVars(double x_init[10]) const
{
  double OB_i[4];
  double Length_init_idx_0;
  double Rot_idx_0;
  double Rot_idx_1;
  double Rot_tmp;
  OB_i[0] = pInitial_[0] - jInitial_[0];
  OB_i[1] = pInitial_[1] - jInitial_[1];
  OB_i[2] = pInitial_[0] - jInitial_[2];
  OB_i[3] = pInitial_[1] - jInitial_[3];
  Rot_tmp = coder::b_norm(*(double(*)[2]) & OB_i[0]);
  Rot_idx_0 = OB_i[0] / Rot_tmp;
  Rot_idx_1 = OB_i[1] / Rot_tmp;
  //  vector / norm (vector)
  Length_init_idx_0 = Rot_tmp;
  //  Gives the length of the Actuator arms
  Rot_tmp = coder::b_norm(*(double(*)[2]) & OB_i[2]);
  //  vector / norm (vector)
  //  Gives the length of the Actuator arms
  x_init[0] = Length_init_idx_0;
  x_init[1] = 0.0;
  x_init[2] = Rot_tmp;
  x_init[3] = 0.0;
  x_init[4] = std::atan2(Rot_idx_1, Rot_idx_0);
  x_init[5] = 0.0;
  x_init[6] = std::atan2(OB_i[3] / Rot_tmp, OB_i[2] / Rot_tmp);
  x_init[7] = 0.0;
  x_init[8] = xTarget_;
  x_init[9] = yTarget_;
}

//
// Arguments    : coder::array<double, 2U> &time_span
// Return Type  : void
//
void CalcSysVars::getTimeSpan(coder::array<double, 2U> &time_span) const
{
  coder::linspace(startTime_, endTime_, (endTime_ - startTime_) / timeSteps_,
                  time_span);
}

//
// Arguments    : const double coordinates[8]
//                const double time_params[3]
// Return Type  : void
//
void CalcSysVars::init(const double coordinates[8], const double time_params[3])
{
  //  Constructor
  pInitial_[0] = coordinates[0];
  pInitial_[1] = coordinates[1];
  jInitial_[0] = coordinates[2];
  jInitial_[1] = coordinates[3];
  jInitial_[2] = coordinates[4];
  jInitial_[3] = coordinates[5];
  xTarget_ = coordinates[6];
  yTarget_ = coordinates[7];
  startTime_ = time_params[0];
  endTime_ = time_params[1];
  timeSteps_ = time_params[2];
}

//
// File trailer for CalcSysVars.cpp
//
// [EOF]
//
